prymer/offtarget/bwa.py

```diff
@@ -275,81 +255,29 @@ def __init__(
-        # -Z = interactive mode (no input buffer and force processing with empty lines between recs)
-        command: list[str] = [
-            f"{executable_path}",
-            "aln",
-            "-t",
-            f"{threads}",
-            "-n",
-            f"{max_mismatches}",
-            "-o",
-            f"{max_gap_opens}",
-            "-e",
-            f"{max_gap_extensions}",
-            "-i",
-            f"{min_indel_to_end_distance}",
-            "-l",
-            f"{seed_length}",
-            "-k",
-            f"{max_mismatches_in_seed}",
-            "-X",
-            f"{max_hits}",
-            "-N",
-            "-S",
-            "-Z",
-            "-D",
-            f"{ref}",
-            "/dev/stdin",
-        ]
+            raise FileNotFoundError(f"{message}\nIndex with: `bwa index {ref}`")
 
-        super().__init__(command=command, stderr=subprocess.PIPE)
-
-        header = []
-        for line in self._subprocess.stdout:
-            if line.startswith("@"):
-                header.append(line)
-            if line.startswith("@PG"):
-                break
-
-        self.header = AlignmentHeader.from_text("".join(header))
-
-        # NB: ExecutableRunner will default to redirecting stderr to /dev/null. However, we would
-        # like to preserve stderr messages from bwa for potential debugging. To do this, we create
-        # a single thread to continuously read from stderr and redirect text lines to a debug
-        # logger. The close() method of this class will additionally join the stderr logging thread.
-        self._logger = logging.getLogger(self.__class__.__qualname__)
-        self._stderr_thread = Thread(
-            daemon=True,
-            target=self._stream_to_sink,
-            args=(self._subprocess.stderr, self._logger.debug),
-        )
-        self._stderr_thread.start()
-
-    def __signal_bwa(self) -> None:
-        """Signals BWA to process the queries."""
-        self._subprocess.stdin.flush()
-        # NB: the executable compiled on different platforms requires a different number of newlines
-        # NB: it is not understood why, but 16 newlines seems to work for all platforms tested
-        self._subprocess.stdin.write("\n" * 16)
-        self._subprocess.stdin.flush()
-
-    @override
-    def close(self) -> bool:
-        """
-        Gracefully terminates the underlying subprocess if it is still running.
+        seq_dict = ref.with_suffix(".dict")
+        with seq_dict.open("r") as fh:
+            with AlignmentFile(fh) as reader:
+                self.header: AlignmentHeader = reader.header
 
-        Returns:
-            True: if the subprocess was terminated successfully
-            False: if the subprocess failed to terminate or was not already running
-        """
-        safely_closed: bool = super().close()
-        self._stderr_thread.join()
-        return safely_closed
+        threads = os.cpu_count() if threads is None else threads
```

Comment:
Something like this will respect more of the OS controls on the running Python process:

```python
def available_cores() -> int:
    """Determine the number of available cores."""
    cores: int = 1
    try:
        from os import sched_getaffinity

        cores = len(sched_getaffinity(0))
    except ImportError:
        import multiprocessing

        cores = multiprocessing.cpu_count()
    except NotImplementedError:
        cores = os.cpu_count() or cores

    return cores
```

The third branch is only necessary if you are supporting older Pythons.

Reply:
Done